Add performance tweaks page (#540)
shopware/docs: guides/hosting/performance/performance-tweaks.md

## Strong CPU

For the server setup, pay special attention to the speed of the CPU. This applies to all servers (app, SQL, Elasticsearch, Redis). Usually, it is more optimal to choose a slightly stronger CPU instead of many weaker ones. But this has to be determined more precisely depending on the project and load. Experience has shown so far that systems with powerful CPUs finish processes faster and can therefore release the resources sooner.

## Redis

We recommend setting up at least four Redis servers for the following resources:
- Session + cart - [Read more](../performance/session.md)
- cache.object - [Read more](../performance/caches.md#example-replace-some-cache-with-redis)
- Lock + Increment storage - [Read more](../performance/increment.md)
- Enqueue - [Read more](../infrastructure/message-queue.md#transport-redis-example)

Instead of setting up a Redis server for `enqueue`, you can also work directly with [RabbitMQ](../infrastructure/message-queue.md#transport-rabbitmq-example)

> The PHP Redis extension should be preferred to be installed on the machine as of Predis to use a persistent Redis connection. Persistent connection opens per FPM process a Redis connection and uses it for all requests, which can help in high load scenarios as each request doesn't have to open and close a connection and may hit the system max open sockets

Reviewer: ```suggestion
The PHP Redis extension provides persistent Redis connections. Persistent connections can help in high load scenarios as each request doesn't have to open and close connections. Using non-persistent Redis connections can also hit the system's maximum of open sockets. Because of these limitations, the Redis extension is preferred over Predis.
```

When a Redis cluster is in usage, the `php.ini` setting `redis.clusters.cache_slots=1` should be set to skip the cluster node lookup on each connection 

## Logging

Set the log level of monolog to `error` to reduce the amount of logged events. Also, limiting the `buffer_size` of monolog prevents memory overflows for long-lived jobs:

```yaml
monolog:
    handlers:
        main:
            level: error  
            buffer_size: 30
        business_event_handler_buffer:
            level: error
```

> The `business_event_handler_buffer` handler logs flows. Setting it to `error` will disable the logging of flow activities that succeed

Reviewer: ```suggestion
The `business_event_handler_buffer` handler logs flows. Setting it to `error` will disable the logging of flow activities that succeed.
```

## Filesystem

> In a multi-app-server system, manage specific directories over a shared filesystem. This includes assets as well as theme files and private and public filesystems. The recommendation is an S3 compatible bucket instead of NFS shares or GCP backend.

Reviewer: Do we prefer S3 over GCP? Or Do we prefer S3 and GCP over NFS? This is unclear from this sentence

[Read more](../infrastructure/filesystem.md)

## HTTP Cache

To ensure a high RPS (requests per second), Shopware offers an integrated HTTP cache with a possible reverse proxy configuration. Any system that handles high user numbers should always use HTTP caching to reduce server resources.

To enable this set `SHOPWARE_HTTP_CACHE_ENABLED=1` in the `.env`

### Reverse Proxy Cache

When you have a lot of app servers, you should consider using a reverse proxy cache like Varnish. Shopware offers default configuration for Varnish out of the box.

[Read more](../infrastructure/reverse-http-cache.md)

### logged-in / cart-filled

By default, Shopware can no longer deliver complete pages from a cache for a logged-in customer or if products are in the shopping cart. As soon as this happens, the user sessions differ, and the context rules could be different depending on the user. The result is different content for each customer; a good example is the [Dynamic Access](https://store.shopware.com/swag140583558965t/dynamic-access.html) plugin.

> However, if the project does not require such functionality, pages can also be cached by the HTTP cache/reverse proxy. The HTTP cache significantly reduces the load for the system. To disable Cache invalidation in these cases:

Reviewer: ```suggestion
However, if the project does not require such functionality, pages can also be cached by the HTTP cache/reverse proxy. To disable Cache invalidation in these cases:
```

```yaml
shopware:
    cache:
        invalidation:
            http_cache: []
```

### Delayed invalidation

> For systems with a high update frequency for the inventory (products, categories), it is recommended to activate a delay for the cache invalidation. Suppose the instruction to delete the cache entries for a specific product or category occurs in the code. In that case, the cache is not deleted directly but processed by a background task afterward. However, if another process also tries to invalidate one of the cache entries of the first process, there will be no overlap, but the timer for the invalidation of this cache entry will only reset.

Reviewer: ```suggestion
A delay for the cache invalidation can be activated for systems with a high update frequency for the inventory (products, categories). Once the instruction to delete the cache entries for a specific product or category occurs, they are not deleted instantly but processed by a background task afterwards. Thus, if two processes invalidate the cache in quick succession, the timer for the invalidation of this cache entry will only reset.
```

When using Elasticsearch, it is important to set the `SHOPWARE_ES_THROW_EXCEPTION=1` `.env` variable. This ensures that if an error occurs when querying the data via Elasticsearch, there is no fallback to the MySQL server. In large projects, failure of the Elasticsearch leads to the MySQL server being completely overloaded otherwise.

[Read more](../infrastructure/elasticsearch/elasticsearch-setup.md)

## Prevent mail data updates
To provide autocompletion for the different mail templates in the administration UI, Shopware has a mechanism that writes an example mail into the database when sending the mail.

With the `shopware.mail.update_mail_variables_on_send` configuration, you can disable this source of database load:

```yaml
shopware:
    mail:
        update_mail_variables_on_send: false
```

[Read more](https://github.com/shopware/platform/blob/trunk/adr/2022-03-25-prevent-mail-updates.md)

## Increment storage

The increment storage is used to store the state and display it in the Administration.

> This storage increments or decrements a given key in a transaction-safe way, which causes locks upon the storage. Therefore, we recommend moving this load to a separate Redis:

Reviewer: ```suggestion
This storage increments or decrements a given key in a transaction-safe way, which causes locks upon the storage. Therefore, we recommend moving this source of server load to a separate Redis:
```

```yaml
shopware:
    increment:
        user_activity:
          type: 'redis'
          config:
            url: 'redis://host:port/dbindex'

        message_queue:
          type: 'redis'
          config:
            url: 'redis://host:port/dbindex'
```

If you don't need such functionality, it is highly recommended to disable this behavior by using `array` as type.

[Read more](../performance/increment.md)

## Lock storage
Shopware uses [Symfony's Lock component](https://symfony.com/doc/5.4/lock.html) to implement locking functionality.
By default, Symfony will use a local file-based lock store, which breaks in multi-machine (cluster) setups. This is avoided by using one of the [supported remote stores](https://symfony.com/doc/5.4/components/lock.html#available-stores).

```yaml
framework:
    lock: 'redis://host:port'
```

[Read more](../performance/lock-store.md)

## Sending mails with the Queue

> Shopware sends the mails by default synchronously. This process can take a while on load or when the remote SMTP server is struggling. For this purpose, it is possible to handle the mails in the message queue. To enable this, add the following config to your config

Reviewer: ```suggestion
Shopware sends the mails by default synchronously. This process can take a while when the remote SMTP server is struggling. For this purpose, it is possible to handle the mails in the message queue. To enable this, add the following config to your config
```
